README.md
# Flextech

Projeto de componentes flutter da empresa Flextech

## Sobre

add no pubspec.yaml a lib flextech

## Componentes



CHANGELOG.md
## [0.0.1] - Inputs

Adicionado TextInput, NumberInput e TextFormInput

## [0.0.2] - Bug Fix
Corrigido bugs

## [0.0.3] - ConstraintsCanvas
adicionado componente ConstraintCanvas


test/flextech_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flextech/flextech.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/inputs/text_input/text_input.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class TextInput extends StatefulWidget {
  final String text;
  final String label;
  final String errorText;
  final ValueChanged<String> onChanged;
  final String hintText;
  final Widget icon;
  final int minLines;
  final int maxLines;
  final int maxLength;
  final bool enabled;
  final bool isPassword;
  final bool autoFocus;
  final FocusNode focusNode;
  final EdgeInsetsGeometry contentPadding;
  final TextEditingController controller;
  final TextInputType keyboardType;
  final TextInputAction textInputAction;
  final List<TextInputFormatter> inputFormatters;
  final Function onSubmited;
  final ThemeData theme;

  const TextInput({
    Key key,
    this.text,
    this.icon,
    this.autoFocus = false,
    this.isPassword = false,
    this.controller,
    this.hintText,
    this.keyboardType,
    this.minLines = 1,
    this.maxLines = 1,
    this.enabled,
    this.contentPadding,
    this.maxLength,
    this.inputFormatters,
    this.errorText,
    @required this.label,
    @required this.theme,
    this.onChanged,
    this.textInputAction,
    this.onSubmited,
    this.focusNode,
  }) : super(key: key);

  @override
  _TextInputState createState() => _TextInputState();
}

class _TextInputState extends State<TextInput> {
  bool focado = false;

  @override
  void initState() {
    focado = widget.autoFocus;

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.theme.colorScheme == null) {
      throw Exception("É necesario implementar o colorSheme no ThemeData");
    }

    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Focus(
        onFocusChange: (hasFocus) {
          setState(() {
            focado = hasFocus;
          });
        },
        child: TextField(
          style: widget.theme.textTheme.headline4
              .copyWith(color: widget.theme.colorScheme.secondary),
          onChanged: widget.onChanged,
          maxLength: widget.maxLength,
          minLines: widget.minLines, //Normal textInputField will be displayed
          maxLines: widget.maxLines, //
          enabled: widget.enabled, //
          obscureText: widget.isPassword,
          controller: widget.controller,
          autofocus: widget.autoFocus,
          textCapitalization: TextCapitalization.words,
          focusNode: widget.focusNode,
          keyboardType: widget.keyboardType,
          textAlign: TextAlign.start,
          textInputAction: widget.textInputAction,
          onSubmitted: widget.onSubmited,
          cursorColor: Colors.white,
          decoration: InputDecoration(
              errorText: (this.widget.errorText == null
                  ? null
                  : this.widget.errorText),
              contentPadding: widget.contentPadding ??
                  EdgeInsets.only(left: 20.0, right: 10),
              border: new OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
                borderRadius: const BorderRadius.all(
                  const Radius.circular(10.0),
                ),
              ),
              enabledBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
              ),
              focusedBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.primary, width: 1.0),
              ),
              focusColor: Colors.red,
              hintStyle: widget.theme.textTheme.headline1.copyWith(
                  fontSize: 14,
                  fontWeight: FontWeight.normal,
                  color: widget.theme.colorScheme.secondary),
              hintText: widget.hintText,
              filled: true,
              labelText: widget.label,
              labelStyle: widget.theme.textTheme.headline1.copyWith(
                  fontSize: 15.0,
                  color: (focado
                      ? widget.theme.colorScheme.primary
                      : widget.theme.colorScheme.secondary)),
              suffixIcon: widget.icon,
              fillColor: Colors.white),
        ),
      ),
    );
  }
}


lib/inputs/text_input/text_form_input.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class TextFormInput extends StatefulWidget {
  final String text;
  final String label;
  final String errorText;
  final ValueChanged<String> onChanged;
  final String hintText;
  final Widget icon;
  final int minLines;
  final int maxLines;
  final int maxLength;
  final bool enabled;
  final bool isPassword;
  final bool autoFocus;
  final FocusNode focusNode;
  final EdgeInsetsGeometry contentPadding;
  final TextEditingController controller;
  final TextInputType keyboardType;
  final TextInputAction textInputAction;
  final List<TextInputFormatter> inputFormatters;
  final Function(String) validator;
  final ThemeData theme;

  const TextFormInput(
      {Key key,
      this.text,
      this.icon,
      this.autoFocus = false,
      this.isPassword = false,
      this.controller,
      this.hintText,
      this.keyboardType,
      this.minLines = 1,
      this.maxLines = 1,
      this.enabled,
      this.contentPadding,
      this.maxLength,
      this.inputFormatters,
      this.errorText,
      @required this.label,
      this.onChanged,
      this.textInputAction,
      this.focusNode,
      @required this.theme,
      this.validator})
      : super(key: key);

  @override
  _TextFormInputState createState() => _TextFormInputState();
}

class _TextFormInputState extends State<TextFormInput> {
  bool focado = false;

  @override
  void initState() {
    focado = widget.autoFocus;

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Focus(
        onFocusChange: (hasFocus) {
          setState(() {
            focado = hasFocus;
          });
        },
        child: TextFormField(
          style: widget.theme.textTheme.headline4
              .copyWith(color: widget.theme.colorScheme.secondary),
          onChanged: widget.onChanged,
          maxLength: widget.maxLength,
          minLines: widget.minLines, //Normal textInputField will be displayed
          maxLines: widget.maxLines, //
          enabled: widget.enabled, //
          validator: widget.validator,
          obscureText: widget.isPassword,
          controller: widget.controller,
          autofocus: widget.autoFocus,
          textCapitalization: TextCapitalization.words,
          focusNode: widget.focusNode,
          keyboardType: widget.keyboardType,
          textAlign: TextAlign.start,
          textInputAction: widget.textInputAction,
          cursorColor: Colors.white,
          decoration: InputDecoration(
              errorText: (this.widget.errorText == null
                  ? null
                  : this.widget.errorText),
              contentPadding: widget.contentPadding ??
                  EdgeInsets.only(left: 20.0, right: 10),
              border: new OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
                borderRadius: const BorderRadius.all(
                  const Radius.circular(10.0),
                ),
              ),
              enabledBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
              ),
              focusedBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.primary, width: 1.0),
              ),
              focusColor: Colors.red,
              hintStyle: widget.theme.textTheme.headline1.copyWith(
                  fontSize: 14,
                  fontWeight: FontWeight.normal,
                  color: widget.theme.colorScheme.secondary),
              hintText: widget.hintText,
              filled: true,
              labelText: widget.label,
              labelStyle: widget.theme.textTheme.headline1.copyWith(
                  fontSize: 15.0,
                  color: (focado
                      ? widget.theme.colorScheme.primary
                      : widget.theme.colorScheme.secondary)),
              suffixIcon: widget.icon,
              fillColor: Colors.white),
        ),
      ),
    );
  }
}


lib/inputs/formatter/numeric_formatter.dart
import 'package:flutter/services.dart';
import 'package:intl/intl.dart';

class NumericFormatter extends TextInputFormatter {
  TextEditingValue formatEditUpdate(
      TextEditingValue oldValue, TextEditingValue newValue) {
    if (newValue.text.length == 0) {
      return newValue.copyWith(text: '');
    } else if (newValue.text.compareTo(oldValue.text) != 0) {
      int selectionIndexFromTheRight =
          newValue.text.length - newValue.selection.end;

      final f = new NumberFormat("#,###", 'es_PY');

      int num = int.parse(newValue.text.replaceAll(f.symbols.GROUP_SEP, ''));
      final newString = f.format(num);
      return new TextEditingValue(
        text: newString,
        selection: TextSelection.collapsed(
            offset: newString.length - selectionIndexFromTheRight),
      );
    } else {
      return newValue;
    }
  }


}


lib/inputs/formatter/numeric_decimal_formatter.dart
import 'dart:math';

import 'package:flutter/services.dart';
import 'package:intl/intl.dart';

/// The `symbol` argument is used to symbol of NumberFormat.
/// Put '\$' for symbol
///
/// The `locale` argument is used to locale of NumberFormat.
/// Put 'en' or 'es' for locale
///
/// The `decimalDigits` argument is used to decimalDigits of NumberFormat.
/// Defaults `decimalDigits` is 2.
class NumericDecimalFormatter extends TextInputFormatter {
  NumericDecimalFormatter({
    this.symbol = '',
    this.locale,
    this.decimalDigits = 2,
  });

  final String symbol;
  final String locale;
  final int decimalDigits;

  @override
  TextEditingValue formatEditUpdate(
      TextEditingValue oldValue, TextEditingValue newValue) {
    final format = NumberFormat.currency(
        locale: locale, decimalDigits: decimalDigits, symbol: symbol);
    String newText = newValue.text.replaceAll(RegExp('[^0-9]'), '');
    if (newText.trim() == '') {
      newValue.copyWith(text: '');
    } else if (newText == '0') {
      newValue.copyWith(text: '');
    } else if (newText == '00') {
      return TextEditingValue(
          text: '', selection: TextSelection.collapsed(offset: 0));
    }
    if (newValue.text.trim() == oldValue.text.trim()) {
      newText = newText.substring(0, newText.length - 1);
    }

    dynamic newInt = int.parse(newText);
    if (decimalDigits > 0) {
      newInt /= pow(10, decimalDigits);
    }
    String newString = format.format(newInt);
    return TextEditingValue(
      text: newString,
      selection: TextSelection.collapsed(offset: newString.length),
    );
  }
}


lib/inputs/number_input/number_input.dart
import 'package:flextech/inputs/formatter/numeric_decimal_formatter.dart';
import 'package:flextech/inputs/formatter/numeric_formatter.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:intl/intl.dart';

class NumberInput extends StatefulWidget {
  final double numberValue;
  final String label;
  final String errorText;
  final ValueChanged<double> onChanged;
  final String hintText;
  final Widget icon;
  final int minLines;
  final int maxLines;
  final int maxLength;
  final bool enabled;
  final bool autoFocus;
  final FocusNode focusNode;
  final EdgeInsetsGeometry contentPadding;
  final TextInputAction textInputAction;
  final List<TextInputFormatter> inputFormatters;
  final Function onSubmited;
  final int precision;
  final ThemeData theme;

  const NumberInput({
    Key key,
    this.numberValue,
    @required this.label,
    this.errorText,
    this.onChanged,
    this.hintText,
    this.icon,
    this.minLines = 1,
    this.maxLines = 1,
    this.maxLength,
    this.enabled,
    this.autoFocus = false,
    @required this.theme,
    this.focusNode,
    this.contentPadding,
    this.textInputAction,
    this.inputFormatters,
    this.onSubmited,
    this.precision = 0,
  }) : super(key: key);

  @override
  _NumberInputState createState() => _NumberInputState();
}

class _NumberInputState extends State<NumberInput> {
  bool focado = false;
  String value = "0";

  @override
  void initState() {
    focado = widget.autoFocus;

    NumberFormat f;

    if (widget.precision == 0) {
      f = new NumberFormat("#,###", 'es_PY');
    } else {
      f = new NumberFormat("#,##0.00", 'es_PY');
    }

    if (widget.numberValue == null) {
      value = "";
    } else {
      value = f.format(widget.numberValue);
    }

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.theme.colorScheme == null) {
      throw Exception("É necesario implementar o colorSheme no ThemeData");
    }
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Focus(
        onFocusChange: (hasFocus) {
          setState(() {
            focado = hasFocus;
          });
        },
        child: TextFormField(
          style: this
              .widget
              .theme
              .textTheme
              .headline4
              .copyWith(color: widget.theme.colorScheme.secondary),
          inputFormatters: [
            (widget.precision == 0
                ? NumericFormatter()
                : NumericDecimalFormatter(
                    decimalDigits: widget.precision, locale: "es_PY"))
          ],
          onChanged: (value) {
            double val = 0;

            if (widget.precision == 0) {
              val = double.parse(value.replaceAll(".", ""));
            } else {
              String s = value.replaceAll(".", "");
              value = s.replaceAll(",", ".");
              val = double.parse(value);
            }

            widget.onChanged(val);
          },
          maxLength: widget.maxLength,
          minLines: widget.minLines, //Normal textInputField will be displayed
          maxLines: widget.maxLines, //
          enabled: widget.enabled, //
          autofocus: widget.autoFocus,
          focusNode: widget.focusNode,
          keyboardType: TextInputType.numberWithOptions(signed: false),
          textAlign: TextAlign.end,
          textInputAction: widget.textInputAction,
          initialValue: value,
          onFieldSubmitted: widget.onSubmited,
          cursorColor: Colors.white,
          decoration: InputDecoration(
              errorText: (this.widget.errorText == null
                  ? null
                  : this.widget.errorText),
              contentPadding: widget.contentPadding ??
                  EdgeInsets.only(left: 20.0, right: 10),
              border: new OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
                borderRadius: const BorderRadius.all(
                  const Radius.circular(10.0),
                ),
              ),
              enabledBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: widget.theme.colorScheme.secondary, width: 1.0),
              ),
              focusedBorder: OutlineInputBorder(
                borderSide: BorderSide(
                    color: this.widget.theme.colorScheme.primary, width: 1.0),
              ),
              hintStyle: this.widget.theme.textTheme.headline1.copyWith(
                  fontSize: 14,
                  fontWeight: FontWeight.normal,
                  color: widget.theme.colorScheme.secondary),
              hintText: widget.hintText,
              filled: true,
              labelText: widget.label,
              labelStyle: this.widget.theme.textTheme.headline1.copyWith(
                  fontSize: 15.0,
                  color: (focado
                      ? this.widget.theme.colorScheme.primary
                      : widget.theme.colorScheme.secondary)),
              suffixIcon: widget.icon,
              fillColor: Colors.white,
              hoverColor: widget.theme.colorScheme.secondary),
        ),
      ),
    );
  }
}
